### PR TITLE
Black Boxes for `toEnum` / `fromEnum`

### LINKED ISSUE #2046: Clash over-eagerly reports "dubious primitive instantiations"
Since https://github.com/clash-lang/clash-compiler/commit/78852675608b9878cb18be560940d7c43805f474, Clash has been more eager to report when designs include primitives on types like `Integer` and `Natural`. These warnings are desirable when such operations remain in the output, but Clash can be too eager to show them to users (see #1918). Before Clash 1.6 we should improve the quality of this reporting to only show such warnings when they are strictly necessary (i.e. when the primitives are contained in the result of normalization).

Designs like this one (from #1918) should not display the "dubious primitive instantiation" warning, as the primitive is not rendered in the resulting HDL.

```haskell
module T2046 where

import Clash.Prelude

{-# NOINLINE topEntity #-}
topEntity
  :: Vec 5 Int
  -> Index 5
  -> Int
topEntity = (!!)
```

### PR DESCRIPTION
**Pull request by alex-mckenna**
This PR introduces black boxes for `toEnum` / `fromEnum` on the numeric types in `Clash.Sized.Internal`, suppressing the seemingly spurious integer black box warnings previously seen in #1918. As an added bonus, a commit is included which propagates the `-Werror` flag from GHC to Clash, allowing primitive warnings to be thrown as `ClashException` when desired.

Fixes #2046

## Still TODO:

  - [x] Write a changelog entry (see changelog/README.md)
  - [x] Check copyright notices are up to date in edited files
  - [x] Add a test based on `toEnum`, the current test uses `fromEnum` only

[comment]: # (Depending on the change, some of these points may not be necessary. If so, leave the boxes unchecked so it is easier for a reviewer to see what has been omitted.)

### CODE REVIEW

File: clash-prelude/src/Clash/Sized/Internal/BitVector.hs [hunk 1]
```diff
@@ -81,6 +83,7 @@ module Clash.Sized.Internal.BitVector
   , gt#
   , le#
     -- *** Enum (not synthesizable)
+  , toEnum#
```
Comment (DigitalBrains1) [inline clash-prelude/src/Clash/Sized/Internal/BitVector.hs:86]: The heading says _not synthesizable_, but `toEnum#` is synthesizable, right? And shouldn't `fromEnum#` be exported as well?

This same thing applies to all following added `toEnum#` and `fromEnum#` functions.

**Review by DigitalBrains1 — CHANGES_REQUESTED** [review]

File: clash-prelude/src/Clash/Sized/Internal/BitVector.hs (hunk 1 above)
Comment (alex-mckenna) [reply clash-prelude/src/Clash/Sized/Internal/BitVector.hs:86]: Yes, and yes. I'm still debugging the testsuite failures though right now :upside_down_face:

Comment (DigitalBrains1) [reply clash-prelude/src/Clash/Sized/Internal/BitVector.hs:86]: I'm not at my sharpest, I kinda missed the _Draft_ status :-).

File: changelog/2022-01-31T16_53_31+01_00_werror_clash.md [hunk 2]
```diff
@@ -0,0 +1 @@
+CHANGED: Clash now respects the -Werror option from GHC
```
Comment (martijnbastiaan) [inline changelog/2022-01-31T16_53_31+01_00_werror_clash.md:1]: :tada:

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]
LGTM, _but_ I'd to see some HDL simulation tests using the blackboxes.

**Review by DigitalBrains1 — APPROVED** [review]

**Comment by alex-mckenna** [comment]
@martijnbastiaan I added a `testBench` to round-trip values through the black boxes. I think this should be enough

File: tests/shouldwork/Issues/T2046BType.hs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module T2046BType where
+
+import Clash.Prelude
+
+type T2046B =
+  ( (Index 1, Index 2)
+  , (Unsigned 1, Unsigned 2)
+  , (Signed 1, Signed 2)
+  , (BitVector 1, BitVector 2)
+  )
```
Comment (alex-mckenna) [inline tests/shouldwork/Issues/T2046BType.hs:10]: Can you test `Bit` too, one of the functions changed for that

Comment (martijnbastiaan) [reply tests/shouldwork/Issues/T2046BType.hs]: It broke CI, good luck.

Comment (martijnbastiaan) [reply tests/shouldwork/Issues/T2046BType.hs]:
```
            iverilog (make testBench):      FAIL (0.04s)
              Program /usr/bin/iverilog failed with error-code 2.
              
              Full invocation:
              
                /usr/bin/iverilog -I T2046BType.o -I T2046B.dyn_hi -I T2046B.hi -I T2046B.testBench -I T2046BType.dyn_o -I T2046BType.dyn_hi -I T2046B.topEntity -I T2046B.dyn_o -I T2046BType.hi -I T2046B.o -g2 -s testBench -o testBench.exe /tmp/clash-test-b57a279493374e3d/T2046B.testBench/testBench.v /tmp/clash-test-b57a279493374e3d/T2046B.topEntity/topEntity.v
                
              
              Stderr was:
              
                /tmp/clash-test-b57a279493374e3d/T2046B.topEntity/topEntity.v:98: syntax error
                /tmp/clash-test-b57a279493374e3d/T2046B.topEntity/topEntity.v:98: error: syntax error in continuous assignment
              
              Stdout was:
```

**Review by martijnbastiaan — APPROVED** [review]

**Pull request merged**